File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.Stack;

public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        Stack<Integer> stack = new Stack<>();
        ArrayList<String> ans = new ArrayList<String>();
        int n = Integer.parseInt(br.readLine());
        int idx=1;
        boolean flag = false;

        for(int i=0; i<n; i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            int num = Integer.parseInt(st.nextToken());
            
            if(idx <= num){
                int t = num - idx;
                for(int j=0; j<=t;j++){
                    stack.push(idx++);
                    ans.add("+");
                }
                stack.pop();
                ans.add("-");
            }
            else{
                if(!stack.isEmpty() && num == stack.peek()){
                    stack.pop();
                    ans.add("-");
                }
                else{
                    flag = true;
                    break;
                }
            }
            
        }
        if(flag) bw.write("NO");
        else{
            for(String str : ans){
                bw.write(str+"\n");
            }
        }
        bw.close();
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.cc
#include <bits/stdc++.h>
using namespace std;

int main(void)
{
  ios::sync_with_stdio(0);
  cin.tie(0);

  stack<int> s;
  vector<char> ans;
  int n,num,s_num=0,flag =0;
  cin>>n;
  s.push(s_num);
  s_num++;
  for (int i=0; i<n; i++){
    cin >> num;
    if (num > s.top()){
        if (num < s_num){
        flag = 1;
        break;
      }
      int cnt = num-s_num+1;
      for (int j=0; j<cnt; j++){
        s.push(s_num);
        s_num++;
        ans.push_back('+');
      }
      s.pop();
      ans.push_back('-');
    }
    else if (num < s.top()){
      while(num != s.top()){
        if (num > s.top()) {
          flag = 1;
          break;
          }
        s.pop();
        ans.push_back('-');
        if (num == s.top()){
          ans.push_back('-');
          s.pop();
          break;
        }
      }
    }
    else{
      s.pop();
      ans.push_back('-');
    }
    if (flag == 1) break;
  }
  if(flag == 1) cout << "NO";
  else {
    for (int i=0; i<ans.size(); i++){
      cout << ans[i] <<"\n"; 
    }
  }
}